[llvm] Introduce callee_type metadata (#87573)
llvm/llvm-project: llvm/docs/CalleeTypeMetadata.rst

====================
Callee Type Metadata
====================

Introduction
============

> This ``!callee_type`` metadata is introduced as part of an ongoing effort to generate a call graph
> section in the object file. The broader design for the call graph section and the compiler flags which
> will enable the feature will be documented as those changes land. The ``!callee_type`` metadata is used
> to identify types of intended callees of indirect call instructions. The ``!callee_type`` metadata is a
> list of one or more ``!type`` metadata objects (See :doc:`TypeMetadata`) with each ``!type`` metadata
> pointing to a callee's :ref:`type identifier
> <calleetype-type-identifier>`.

Reviewer: ```suggestion
This ``!callee_type`` metadata is introduced to support the generation of a call graph
section in the object file.  The ``!callee_type`` metadata is used
to identify the types of the intended callees of indirect call instructions. The ``!callee_type`` metadata is a
list of one or more ``!type`` metadata objects (See :doc:`TypeMetadata`) with each ``!type`` metadata
pointing to a callee's :ref:`type identifier
<calleetype-type-identifier>`.
```

.. _calleetype-type-identifier:

Type identifier
================

The type for an indirect call target is the callee's function signature.
Mapping from a type to an identifier is an ABI detail.
In the current implementation, an identifier of type T is

> computed as follows:

Reviewer: It may be worth mentioning that this is the same type identifier used by CFI.
